src/query/api/v1/middleware/rewrite.go

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2021  Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package middleware
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/prometheus/prometheus/promql/parser"
+	"go.uber.org/zap"
+
+	"github.com/m3db/m3/src/query/api/v1/handler/prometheus"
+	"github.com/m3db/m3/src/query/api/v1/handler/prometheus/handleroptions"
+	"github.com/m3db/m3/src/query/storage"
+)
+
+// PrometheusRangeRewriteOptions are the options for the prometheus range rewriting middleware.
+type PrometheusRangeRewriteOptions struct {
+	Enabled              bool
+	FetchOptionsBuilder  handleroptions.FetchOptionsBuilder
+	ResolutionMultiplier int
+	Storage              storage.Storage
+}
+
+// PrometheusRangeRewrite is middleware that, when enabled, will rewrite the query parameter
+// on the request (url and body, if present) if it's determined that the query contains a
+// range duration that is less than the resolution that will be used to serve the request.
+// With this middleware disabled, queries like this will return no data. When enabled, the
+// range is updated to be the namespace resolution * a configurable multiple which should allow
+// for data to be returned.
+func PrometheusRangeRewrite(opts Options) mux.MiddlewareFunc {
+	return func(base http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			var (
+				mwOpts   = opts.PrometheusRangeRewrite
+				mult     = mwOpts.ResolutionMultiplier
+				disabled = !mwOpts.Enabled
+			)
+			if disabled || mult == 0 {
+				base.ServeHTTP(w, r)
+				return
+			}
+
+			logger := opts.InstrumentOpts.Logger()
+			if err := rewriteRangeDuration(r, mwOpts, logger); err != nil {
+				logger.Warn("could not rewrite range. "+
+					"continuing with request", zap.Error(err))
```

Comment:
I'd argue this should return an error for the entire query. It seems like there would need to be a code bug for this to return an error and I think it silently disabling during a new release and us not noticing the logs would not be great.

Reply:
Maybe make this configurable and default it to returning a hard error? That way you can always change the behavior temporarily as a get out of jail free card. (although honestly you could just disable it if it was actually causing issues, so maybe not necessary and just to default this to return a hard error)

Reply:
Yeah, will just make this a hard error and if it's problematic someone can just disable the middleware